[Inductor] Break the loop fusion when node2 depends on node1 mutations

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109172

**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

leslie-fang-intel committed:
Break the loop fusion and node2 depends on node1 mutations

[ghstack-poisoned]

jgong5 reviewed (changes requested):
Seems there is test failure...

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel commented:
The PreCI has been passed. @jgong5 please help to take a look again.

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -1485,6 +1485,16 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
+            # If any buffer used by node2 is a mutation of node1, disable the fusion of node1 and node2
+            for key in node2._body.reads_name2expr.keys():
+                if key in node1.get_mutations():
+                    return False
```
This seems like a heavy-handed solution. Shouldn't we look into why the CSE load cache wasn't invalidated by the mutation of `buf3`?

peterbell10 replied:
Looks like it's because of `mode="atomic_add"`. I guess we need to track `accumulation` dependencies separately from normal write dependencies in the scheduler.

jansel replied:
I agree, let's only do this for atomics not all mutations.

leslie-fang-intel replied:
> Looks like it's because of `mode="atomic_add"`. I guess we need to track `accumulation` dependencies separately from normal write dependencies in the scheduler.

@peterbell10 @jansel Thanks for your suggestions. Did you mean we keep the node fusion and change the `CSE load store cache` with `atomic_add` in the kernel codegen? I draft a version, could you help to take a look?

leslie-fang-intel replied:
> Looks like it's because of `mode="atomic_add"`. I guess we need to track `accumulation` dependencies separately from normal write dependencies in the scheduler.

Thanks for the suggestions. @peterbell10 @jansel @jgong5, please help to take a look again.

- Since the store is with `mode="atomic_add"`, we can't store it into ` CSE store cache` and use it later directly.
- Add the check `mode="atomic_add"` in the `scheduler.py`, so it will break the fusion only with atomics but not all mutations. The basic logic is that to check the `mode` value in all the `store` node and break if `True`.

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/scheduler.py:
```diff
@@ -1485,6 +1485,40 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
         if node2.get_names() & node1.recursive_predecessors:
             return False  # node2 must go before node1
 
+        # If node1 is a FusedSchedulerNode which assumes already has fused several scheduler nodes,
+        # we only check the last scheduler node among the FusedSchedulerNode.
+        check_atomic_add_mutation_node1 = (
+            node1.get_nodes()[-1] if isinstance(node1, FusedSchedulerNode) else node1
+        )
+        if (
+            isinstance(check_atomic_add_mutation_node1, SchedulerNode)
+            and isinstance(node2, SchedulerNode)
+            and isinstance(check_atomic_add_mutation_node1._body, ir.LoopBody)
+            and isinstance(node2._body, ir.LoopBody)
+        ):
+            # Fix Issue: https://github.com/pytorch/pytorch/issues/108963
+            # Disable the fusion of node1 and node2, if:
+            # * Any buffer used by node2 is a mutation of node1
+            # * Store Buffer of node1 is produced by mode of atomic_add
+
+            def _store_as_atomic_add(check_node):
+                for node in check_node._body.get_nodes():
+                    if node.op == "call_method" and node.target == "store":
+                        # Check the mode parameter of store node is atomic_add
+                        if (
+                            "mode" in node.kwargs
+                            and node.kwargs["mode"] == "atomic_add"
+                        ) or (len(node.args) == 5 and node.args[4] == "atomic_add"):
+                            return True
+                return False
```
This seems a bit awkward.  It would be cleaner to add a `node.has_atomic_add(buffer_name)` method.

Or it might be easier to just add a mode field to the dependency object so we don't need to parse the IR.

leslie-fang-intel replied:
Thanks for the suggestion.

- I modify the implementation with option 1: add the method of `has_atomic_add()` for `SchedulerNode` since it looks like a straight-forward change with current implementation. BTW: not sure if my understanding is right, I suppose `SchedulerNode` will has only 1 store buffer, so we don't need to pass in parameter of `buffer_name`.
- For option 2: `add a mode field to the dependency object`. If it looks more cleaner from your point-of-view, could you kindly help with more inputs? I am not sure how to implement it in this option yet.

jansel replied:
FusedSchedulerNode can have many different store buffers.  

For the second one, you would need to modify this:
https://github.com/pytorch/pytorch/blob/2c1554a0323107d821be3ff13df7833b9f0b960d/torch/_inductor/dependencies.py#L254-L256

To record the mode.  Plus some related function in that file for copying/normalizing deps.

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/scheduler.py:
```diff
@@ -681,6 +681,17 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
         return False
 
+    @cache_on_self
+    def has_atomic_add(self):
```
To flesh out my earlier comment a bit more:
> I guess we need to track `accumulation` dependencies separately from normal write dependencies in the scheduler.

Currently in `dependencies.py` we extract `reads` and `writes` for each node and vertical fusion happens when one node reads what the other writes. If we track `mode="atomic_add"` stores separately from writes, then reading from an accumulated value won't be scored as a positive for fusion in the first place. We would also have more information on the accumulation dependency, so we could also say that a node that has an atomic add _can_ be fused with other nodes, but only if they have no other dependencies involving that accumulated buffer. This would allow more fusions than your current code.

jansel reviewed (changes requested):
I think has_atomic_add() is too general because the atomic add could be to an unrelated buffer and this will false-positive.

Either need to check that the buffer names match or modify the dependency analysis here:
https://github.com/pytorch/pytorch/blob/2c1554a0323107d821be3ff13df7833b9f0b960d/torch/_inductor/dependencies.py#L254-L256

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/scheduler.py:
```diff
@@ -1485,6 +1496,32 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
+                            node2_used_buf, sub_schedule_node1
+                        )
+                        for sub_schedule_node1 in node1.get_nodes()
+                    )
```
Thanks and add the `has_atomic_add(buf_name)` for `FusedSchedulerNode`.

leslie-fang-intel commented:
> I think has_atomic_add() is too general because the atomic add could be to an unrelated buffer and this will false-positive.
> 
> Either need to check that the buffer names match or modify the dependency analysis here:

Add the check of buffer names in `has_atomic_add()` which will accept extra input of `check_buf`.

jgong5 commented on torch/_inductor/scheduler.py:
```diff
@@ -785,6 +799,17 @@ def op_counts(self):
         return op_counts
 
+    def has_atomic_add(self, check_buf):
+        if any(
+            (
+                isinstance(sub_schedule_node1, SchedulerNode)
+                and sub_schedule_node1.has_atomic_add(check_buf)
+            )
+            for sub_schedule_node1 in self.get_nodes()
+        ):
+            return True
+        return False
```
```suggestion
        return any(
            (
                isinstance(sub_schedule_node1, SchedulerNode)
                and sub_schedule_node1.has_atomic_add(check_buf)
            )
            for sub_schedule_node1 in self.get_nodes()
        )
```

leslie-fang-intel replied:
Thanks and changed.

jgong5 commented on torch/_inductor/scheduler.py:
```diff
@@ -681,6 +681,20 @@ def can_inplace(self, read_dep: dependencies.MemoryDep):
 
+    def has_atomic_add(self, check_buf):
+        # Return True if check_buf is stored by atomic_add mode
+        for node in self._body.get_nodes():
```
is it better to build a cached helper method that returns the atomic_add buffer so that here we just check if `check_buf` equals to it? This way, you don't have to scan the body IR again and again.

leslie-fang-intel replied:
Thanks for the suggestion and changed. Yean, it's better and can save redundant body IR scan.

leslie-fang-intel committed:
Update on "[Inductor] Break the loop fusion when node2 depends on node1 mutations"


**Summary**
Fix the issue: https://github.com/pytorch/pytorch/issues/108963. After this PR, loop fusion should break when node2 depends on node1's buffer mutation. Take the UT as example:

- Before this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

- After this PR, the generated code is:
```
cpp_fused_div_index_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
    }
    {
        auto tmp0 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<double>(2.0);
        auto tmp2 = tmp0 / tmp1;
        out_ptr1[static_cast<long>(0L)] = tmp2;
    }
}
''')
```


**Test Plan**
```
python -u -m pytest -s -v test_torchinductor.py -k test_mutations_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
